lib/filterx/expr-literal.c

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (c) 2023 Balazs Scheidler <balazs.scheidler@axoflow.com>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "filterx/filterx-expr.h"
+
+typedef struct _FilterXLiteral
+{
+  FilterXExpr super;
+  FilterXObject *object;
+} FilterXLiteral;
+
+static FilterXObject *
+_eval(FilterXExpr *s)
+{
+  FilterXLiteral *self = (FilterXLiteral *) s;
+  return filterx_object_ref(self->object);
```

Comment:
This seems a bit incorrect. Each evaluation creates a new reference, but we only have a single `unref` when freeing this literal. If I understand correctly, the current implementation of the freezing mechanism hides the issue with this asymmetry, but I think it would be worth it to fix it anyway.

Reply:
eval always returns a reference and it is the responsibility of the caller to free this reference.

For frozen objects, both _ref and _unref are actually noops. For non-frozen objects the _ref is needed and unref is done by the caller for every eval() call.